requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,600 @@
+}
+```
+
+Adding a few filters
+
+```json
+{
+  "type": "events",
+  "filters": {
+    "event": "$pageview",
+    "properties": [{ "key": "$browser", "equals": "Chrome" }],
+    "date_from": "-14d",
+    "date_to": null
+  }
+}
+```
+
+```ts
+interface EventsNode extends Node {
+  type: "events";
+  filters: EventFilters;
+}
+```
+
+This query can be used as input wherever an `EventsNode` is expected. If it's at the root of a query, the frontend will
+receive an `EventsResponse`,  
+
+```ts
+interface EventsResponse {
+  type: "events";
+  query: EventsNode,
+  response: { count: number, results: EventResult[], }
+}
+```
+
+and will likely render the "events list" scene.
+
+If the result comes back with `type: "funnel"`, we'll instead render an insight with the funnel:
+
+```json
+{
+  "type": "funnel",
+  "steps": [
+    {
+      "type": "events"
+    },
+    {
+      "type": "events",
+      "filters": {
+        "event": "$pageview",
+        "properties": [{ "key": "$browser", "equals": "Chrome" }]
+      }
+    }
+  ]
+}
+```
+
+```ts
+interface FunnelNode extends Node {
+  type: "funnel";
+  steps: EventsNode[];
+}
+```
+
+If we want people who dropped off at a step, broken down further, it'd be something like:
+
+```json
+{
+  "type": "persons",
+  "breakdown": ["person.account_type"],
+  "from": {
+    "type": "funnel-dropoff",
+    "step": "2",
+    "from": {
+      "type": "funnels",
+      "steps": [
+        {
+          "type": "events"
+        },
+        {
+          "type": "events",
+          "filters": {
+            "event": "$pageview",
+            "properties": [
+              {
+                "key": "$browser",
+                "equals": "Chrome"
+              }
+            ]
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+```ts
+interface PersonsNode extends Node {
+  type: "persons";
+  breakdown?: string[];
+  from?: FunnelDropoffNode | EventsNode;
+}
+
+interface FunnelDropoffNode extends Node {
+  type: "funnel-dropoff";
+  step: number;
+  from: FunnelNode;
+}
+```
+
+I'm not sure how many levels of abstractions this requires. It'll depend on how this maps to actual SQL.
+
+You could get correlation analysis for a funnel in a similar way:
+
+```json
+{
+  "type": "correlation",
+  "correlation": {
+    "type": "properties",
+    "exclude_names": [
+      "$initial_geoip_postal_code",
+      "$initial_geoip_latitude",
+      "$initial_geoip_longitude"
+    ],
+    "names": ["$all"]
+  },
+  "from": {
+    "type": "funnels",
+    "steps": [
+      {
+        "type": "events"
+      },
+      {
+        "type": "events",
+        "filters": {
+          "event": "$pageview",
+          "properties": [{ "key": "$browser", "equals": "Chrome" }]
+        }
+      }
+    ]
+  }
+}
+```
+
+```ts
+interface CorrelationNode extends Node {
+  type: "correlation";
+  correlation?: CorrelationOptions;
+  from?: FunnelNode;
+}
+```
+
+## Backend
+
+TODO: what is needed in the backend to make this possible?
```

Comment:
Drive by thought: The nested filter is already dramatically better than our current filter monster. 

I think the crux of the complexity is how we'll be handling queries of arbitrary matchups of concepts such as correlation of trends or breakdown on paths? Making sure that we can support arbitrary complexity in querying. Part of the reason the insights are so bespoke is because the queries can end up being formulated very differently. Really interested to hear the ideas on how to make queries composable as I think that will inform the feasibility of nested filters

Reply:
I dug into our queries, and basically all of them follow the same, sort of map-reduce-reduce format:
1. Select all events that are relevant
    - Selects timestamp, aggregation target (person_id, distinct_id) + ‘other’
        - Funnel strict: step_{0,1,…}, latest_{0,1,…} (timestamp), prop_vals (breakdown)
        - Trends: breakdown_value
        - Lifecycle: created_at
        - Paths: path_item
        - Retention: breakdown_values (though this is breadth first)
            - Retention has two: target_event_query and returning_event_query
        - Stickiness: nothing
    - Normally grouped by those selects
2. Reduce to aggregation target, time period + ‘other’
    - Funnel strict: step_(1,2,3)_average_conversion_time_inner, prop (for breakdown)
    - Funnel trends: steps completed
    - Trends: breakdown_value, session_duration
    - Lifecycle: status
    - Path: path_time
    - Stickiness: num_intervals
3. Reduce to counts for the graph

Right now 2 and 3 are often completely intertwined. If we separate them it should be relatively easy to do this. For example:

**Get a list of people at a specific funnel**
The example on line 411 would turn into:
`Select * from persons where person_id IN (select person_id from (FUNNEL OUTPUT STEP 2) where step_1 = 1 and prop = ‘$browser’)`

**Trends query on anyone that dropped off on step 2**
`Select count(1), timestamp from events where person_id IN (select person_id from (FUNNEL OUTPUT STEP 2) where step_1 = 1`

Reply:
> Right now 2 and 3 are often completely intertwined. If we separate them it should be relatively easy to do this. For example:

I might be misunderstanding the proposal but in ClickHouse queries it's often important to combine the two for performance reasons. Otherwise you can end up with badly performing queries due to streaming unneeded data over the network or getting bitten by the lack of query optimizer in clickhouse.

For some examples of this, check out these nifty handbook pages:
- https://posthog.com/handbook/engineering/clickhouse/data-storage#no-query-planner
- https://posthog.com/handbook/engineering/clickhouse/replication#example-query-limit-filter-and-aggregate

Reply:
> We'd still always execute only one query.

I was assuming that. My point is that a flat query like `select sum(...) from events where time filter group by foobar` is going to perform better than semantically identical SQL in 3 layers like:

```sql
select sum(...)
from (
  select some_cols from (
     select * from events where time filter
  )
)
group by foobar
```

We have competing pressures here: for code simplicity we want the nested variant. For performance, we need bespoke code that generates a more optimal query..

Reply:
@macobo hit the nail on the head with regards to the crux of the issue here. CH does not have a cost based query planner and so the structure of your query is _critical_. It determines how the query will be executed (unlike most query engines out there) and will be one of the most challenging parts of implementing this in a way that is optimized for performance.

Reply:
All fair points. I think the principle of the separation of queries into map-reduce-reduce still stands even if we do it within the same (sub)query. We can prototype something but I think this'll be possible.

Reply:
This is useful to know. I do agree we should still be able to reuse the structures more, even with flat queries. I'll be hacking on this next week.

Reply:
If we think of this as an AST, we can solve the problem the same way compilers optimize code: with clever hacks and shortcuts.

For example, taking the "[line 411](https://github.com/PostHog/meta/pull/66/files#diff-2200663423f552a421711aa0cda6877a88fda13dfc78542dd45bf5dfd0a5b4b4R411-R440)" query Tim referred to:

```json
{
  "type": "persons",
  "breakdown": ["person.account_type"],
  "from": {
    "type": "funnel-dropoff",
    "step": "2",
    "from": {
      "type": "funnels",
      "steps": [
        {
          "type": "events"
        },
        {
          "type": "events",
          "filters": {
            "event": "$pageview",
            "properties": [
              {
                "key": "$browser",
                "equals": "Chrome"
              }
            ]
          }
        }
      ]
    }
  }
}
```

All these types, `persons`, `funnel-dropdown`, `funnels`, `events` could evaluate to their own subqueries that can literally be nested within SQL.

However we can run an optimizer as a first pass.

It could replace a tree like `persons -> funnelDropoff -> funnels -> events` with a more optimized `personsFunnelDropOff -> funnelFromEvents` tree if one is available. In this case converting 4 nested sql queries into two. I'm really not sure what's the right level of verbosity here, and I guess that'll only be determined when actually implementing this.